rust/src/operations/restore.rs

```diff
@@ -0,0 +1,288 @@
+//! in file system.
+//! 6) Commit Protocol, all RemoveFile and AddFile actions
+//! into delta log using `try_commit_transaction` (commit will be failed in case of parallel transaction)
+//! 7) If table was modified in parallel then ignore restore and raise exception.
+//!
+//! # Example
+//! ```rust ignore
+//! let table = open_table("../path/to/table")?;
+//! let (table, metrics) = RestoreBuilder::new(table.object_store(), table.state).restore_to_version(1).await?;
+//! ````
+
+use std::cmp::max;
+use std::collections::HashSet;
+use std::ops::BitXor;
+use std::time::{SystemTime, UNIX_EPOCH};
+
+use chrono::{DateTime, Utc};
+use futures::future::BoxFuture;
+use object_store::path::Path;
+use object_store::ObjectStore;
+
+use crate::action::{Action, Add, DeltaOperation, Remove};
+use crate::operations::transaction::{prepare_commit, try_commit_transaction, TransactionError};
+use crate::storage::ObjectStoreRef;
+use crate::table_state::DeltaTableState;
+use crate::{action, DeltaResult, DeltaTable, DeltaTableConfig, DeltaTableError, ObjectStoreError};
+
+/// Errors that can occur during restore
+#[derive(thiserror::Error, Debug)]
+enum RestoreError {
+    #[error("Either the version or datetime should be provided for restore")]
+    InvalidRestoreParameter,
+
+    #[error("Version to restore {0} should be less then last available version {1}.")]
+    TooLargeRestoreVersion(i64, i64),
+
+    #[error("Find missing file {0} when restore.")]
+    MissingDataFile(String),
+}
+
+impl From<RestoreError> for DeltaTableError {
+    fn from(err: RestoreError) -> Self {
+        DeltaTableError::GenericError {
+            source: Box::new(err),
+        }
+    }
+}
+
+/// Metrics from Restore
+#[derive(Default, Debug)]
+pub struct RestoreMetrics {
+    /// Number of files removed
+    pub num_removed_file: usize,
+    /// Number of files restored
+    pub num_restored_file: usize,
+}
+
+/// Restore a Delta table with given version
+/// See this module's documentation for more information
+pub struct RestoreBuilder {
+    /// A snapshot of the to-be-restored table's state
+    snapshot: DeltaTableState,
+    /// Delta object store for handling data files
+    store: ObjectStoreRef,
+    /// Version to restore
+    version_to_restore: Option<i64>,
+    /// Datetime to restore
+    datetime_to_restore: Option<DateTime<Utc>>,
+    /// Ignore missing files
+    ignore_missing_files: bool,
+    /// Protocol downgrade allowed
+    protocol_downgrade_allowed: bool,
+}
+
+impl RestoreBuilder {
+    /// Create a new [`RestoreBuilder`]
+    pub fn new(store: ObjectStoreRef, snapshot: DeltaTableState) -> Self {
+        Self {
+            snapshot,
+            store,
+            version_to_restore: None,
+            datetime_to_restore: None,
+            ignore_missing_files: false,
+            protocol_downgrade_allowed: false,
+        }
+    }
+
+    /// Set the version to restore
+    pub fn restore_to_version(mut self, version: i64) -> Self {
```

Comment:
For consisitency with other builders it would be great to name the methods to configure options `with_*`

Reply:
done